docs/custom-rules.rst

```diff
@@ -0,0 +1,92 @@
+************
+Custom Rules
+************
+
+Creating Custom Rules
+---------------------
+
+Rules are described using a class file per rule. Default rules are named *DeprecatedVariableRule.py*, etc.
+
+Each rule definition should have the following:
+
+* ID: A unique identifier
+* Short description: Brief description of the rule
+* Description: Behaviour the rule is looking for
+* Tags: one or more tags that may be used to include or exclude the rule
+* At least one of the following methods:
+
+  * ``match`` that takes a line and returns None or False, if the line doesn't match the test, and True or a custom message, when it does. (This allows one rule to test multiple behaviours - see e.g. the *CommandsInsteadOfModulesRule*.)
+  * ``matchtask`` that operates on a single task or handler, such that tasks get standardized to always contain a *module* key and *module_arguments* key. Other common task modifiers, such as *when*, *with_items*, etc., are also available as keys, if present in the task.
+
+An example rule using ``match`` is:
+
+.. code-block:: python
+
+    from ansiblelint.rules import AnsibleLintRule
+
+    class DeprecatedVariableRule(AnsibleLintRule):
+
+        id = 'EXAMPLE002'
+        shortdesc = 'Deprecated variable declarations'
+        description = 'Check for lines that have old style ${var} ' + \
+                      'declarations'
+        tags = { 'deprecations' }
+
+        def match(self, file, line):
+            return '${' in line
+
+An example rule using ``matchtask`` is:
+
+.. code-block:: python
+
+    import ansiblelint.utils
+    from ansiblelint.rules import AnsibleLintRule
+
+    class TaskHasTag(AnsibleLintRule):
+        id = 'EXAMPLE001'
+        shortdesc = 'Tasks must have tag'
+        description = 'Tasks must have tag'
+        tags = ['productivity']
+
+        def matchtask(self, file, task):
+            # If the task include another task or make the playbook fail
+            # Don't force to have a tag
+            if not set(task.keys()).isdisjoint(['include','fail']):
+                return False
+
+            # Task should have tags
+            if not task.has_key('tags'):
+                  return True
+
+        return False
```

Comment:
Check the indentation on this line
```suggestion
            return False
```

Reply:
Good catch! It was broken in the original snippet too.